command.go

```diff
@@ -4168,3 +4168,91 @@ func (cmd *LCSCmd) readPosition(rd *proto.Reader) (pos LCSPosition, err error) {
 
 	return pos, nil
 }
+
+// ------------------------------------------------------------------------
+
+type KeyFlags struct {
+	Key   string
+	Flags []string
+}
+
+type KeyFlagsCmd struct {
+	baseCmd
+
+	val []KeyFlags
+}
+
+var _ Cmder = (*KeyFlagsCmd)(nil)
+
+func NewKeyFlagsCmd(ctx context.Context, args ...interface{}) *KeyFlagsCmd {
+	return &KeyFlagsCmd{
+		baseCmd: baseCmd{
+			ctx:  ctx,
+			args: args,
+		},
+	}
+}
+
+func (cmd *KeyFlagsCmd) SetVal(val []KeyFlags) {
+	cmd.val = val
+}
+
+func (cmd *KeyFlagsCmd) Val() []KeyFlags {
+	return cmd.val
+}
+
+func (cmd *KeyFlagsCmd) Result() ([]KeyFlags, error) {
+	return cmd.val, cmd.err
+}
+
+func (cmd *KeyFlagsCmd) String() string {
+	return cmdString(cmd, cmd.val)
+}
+
+func (cmd *KeyFlagsCmd) readReply(rd *proto.Reader) error {
+	n, err := rd.ReadArrayLen()
+	if err != nil {
+		return err
+	}
+
+	if n == 0 {
+		cmd.val = make([]KeyFlags, 0)
+		return nil
+	}
+
+	typ, err := rd.PeekReplyType()
+	if err != nil {
+		return err
+	}
+	array := typ == proto.RespArray
+
+	if array {
+		cmd.val = make([]KeyFlags, n)
+	} else {
+		cmd.val = make([]KeyFlags, n/2)
+	}
```

Comment:
I don't understand what's going on here. Why do we need to check the type of response in Redis? Does Redis respond with multiple data types?

Reply:
As you can see https://github.com/redis/go-redis/blob/6790337e5ddff23427a083cbca1140e039bc693b/command.go#L1556
Similarly I've done this for ZSliceCmd as well as this
I'm not sure in which scenario it happens
Do you suggest we should directly make the array without checking the type of response we are getting because the protocol is fixed?

Reply:
This is done to be compatible with both the RESP2 and RESP3 protocols. In older versions of Redis, map type responses are represented as array types, so we need to ensure that they continue to work properly in RESP2.

 There is no such situation with command-getkeys.